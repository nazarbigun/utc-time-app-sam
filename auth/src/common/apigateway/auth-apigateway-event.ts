export interface AuthApigatewayEvent {
    type: 'TOKEN';
    methodArn: string;
    authorizationToken: string;
}
