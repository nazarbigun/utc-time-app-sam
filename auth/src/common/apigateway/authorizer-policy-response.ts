interface PolicyStatement {
    Action: string;
    Effect: string;
    Resource: string;
}

interface PolicyDocument {
    Version: string;
    Statement: PolicyStatement[];
}

export interface AuthorizerPolicyResponse {
    principalId: string;
    policyDocument: PolicyDocument
}
