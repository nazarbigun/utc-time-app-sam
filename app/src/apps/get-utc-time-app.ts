import { ApiGatewayResponse } from '../common/apigateway/apigateway-response';

import { LambdaApp } from './lambda-app';

/**
 * GetUtcTimeApp is a LambdaApp that returns the current utc time.
 */
export class GetUtcTimeApp implements LambdaApp {
    async run(): Promise<ApiGatewayResponse> {
        try {
            const now = new Date();
            const time = `${now.getUTCHours()}:${now.getUTCMinutes()}:${now.getUTCSeconds()}`;

            return { statusCode: 200, body: JSON.stringify({ time }) };

        } catch(err) {
            return { statusCode: 500 };
        }
    }
}
