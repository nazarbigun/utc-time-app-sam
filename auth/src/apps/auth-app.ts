import { SecretsManager, GetSecretValueCommand } from '@aws-sdk/client-secrets-manager';
import { AuthApigatewayEvent } from '../common/apigateway/auth-apigateway-event';
import { AuthorizerPolicyResponse } from '../common/apigateway/authorizer-policy-response';

export class AuthApp {
  async run(event: AuthApigatewayEvent): Promise<AuthorizerPolicyResponse> {
    const authHeader = event.authorizationToken;
    try {
      const client = new SecretsManager({
        region: process.env.AWS_REGION // Your region
      });

      const command = new GetSecretValueCommand({ SecretId: process.env.SECRET_NAME });
      const data = await client.send(command);

      if (authHeader !== data.SecretString) {
        return this.generatePolicy('user', 'Deny', event.methodArn);
      }

      return this.generatePolicy('user', 'Allow', event.methodArn);
    }
    catch (err) {
      console.log(err);
      throw new Error('You have issue with getting secret.')
    }
  }

  private generatePolicy(principalId: string, effect: string, resource: string): AuthorizerPolicyResponse {
    return {
      principalId,
      policyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'execute-api:Invoke',
            Effect: effect,
            Resource: resource
          }
        ]
      }
    };
  }
}
