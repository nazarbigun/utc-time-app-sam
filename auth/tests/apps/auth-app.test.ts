import { AuthApp } from '../../src/apps/auth-app';
import { AuthorizerPolicyResponse } from '../../src/common/apigateway/authorizer-policy-response';
import { AuthApigatewayEventMock } from '../mocks/apigateway-event-mock';

class SecretsManagerMock {
  send = jest.fn().mockReturnValue({ SecretString: 'validToken' })
}

class GetSecretValueCommandMock {}

jest.mock('@aws-sdk/client-secrets-manager', () => {
  return {
    SecretsManager: jest.fn(() => new SecretsManagerMock()),
    GetSecretValueCommand: jest.fn(() => new GetSecretValueCommandMock())
  }
})

describe('GetUtcTimeApp instance', () => {
  describe('run', () => {
    it('should return authorized policy', async () => {
      const app = new AuthApp();
      const response: AuthorizerPolicyResponse = await app.run(new AuthApigatewayEventMock('validToken'));

      expect(response).toMatchSnapshot();
    });

    it('should return not authorized policy', async () => {
      const app = new AuthApp();
      const response: AuthorizerPolicyResponse = await app.run(new AuthApigatewayEventMock('invalidToken'));

      expect(response).toMatchSnapshot();
    });
  });
});
