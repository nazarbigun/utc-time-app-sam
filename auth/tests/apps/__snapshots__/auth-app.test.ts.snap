// Jest Snapshot v1, https://goo.gl/fbAQLP
exports[`GetUtcTimeApp instance run should return authorized policy 1`] = `
Object {
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Allow",
        "Resource": "arn:aws:resource:region:user_id:key:some_id",
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "user",
}
`;

exports[`GetUtcTimeApp instance run should return not authorized policy 1`] = `
Object {
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Deny",
        "Resource": "arn:aws:resource:region:user_id:key:some_id",
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "user",
}
`;
